### PR TITLE
Add significantly more debug logging to UIAHandler

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
Although NVDA has a specific uIA debug logging category, this does not really enable all that many more log messages a part from some debug warnings. thus it is almost useless to ask for a user's log with this flag enabled.

 
### Description of user facing changes
Setting NVDA's log level to debug, and enabling the UIA debug logging category in NVDA's advanced settings panel, will now produce significantly more logging for UIA. This includes logging when all UIA event handler methods are called, along with their parameters, and logging for utility utility methods such as checking if a window or element is native UIA.
 
### Description of development approach
Added a bunch of methods to UIAHandler for generating a string representation for debugging, for things such as UIA elements, property IDs, event IDs, notification types etc.
Added conditional debug logging to all UIA event handler methods, plus isUIAWindow, isNativeUIAElement and getNearestWindowhandle.
 
### Testing strategy:
Ran NVDA for a day or so, performing general tasks, using multiple apps that are native UIA.
Used the new debug logging to aide in investigating QT issues.

### Known issues with pull request:
None known.

### Change log entries:
New features
Changes
Bug fixes
For Developers
* the UIA debug logging category when enabled now produces significantly more logging for UIA event handlers and utilities.
### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by josephsl** [comment]
Subscribing as it will impact Event Tracker add-on once this PR is merged (if merged, I will declare Event Tracker as a legacy add-on). Thanks.

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -1292,6 +1302,11 @@ def __init__(self,windowHandle=None,UIAElement=None,initialUIACachedPropertyIDs=
 		UIACachedWindowHandle=UIAElement.cachedNativeWindowHandle
 		self.UIAIsWindowElement=bool(UIACachedWindowHandle)
 		if not windowHandle:
+			if UIAHandler._isDebug():
+				log.debug(
+					f"No windowHandle for UIA NvDAObject. "
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/__init__.py:1307]: Fix capitalization of NVDA

```suggestion
					f"No windowHandle for UIA NVDAObject. "
```

File: source/UIAHandler/__init__.py [hunk 2]
```diff
@@ -49,6 +49,14 @@
 from . import remote as UIARemote
 
 
+baseCachePropertyIDs = set([
+	UIA.UIA_FrameworkIdPropertyId, UIA.UIA_AutomationIdPropertyId, UIA.UIA_ClassNamePropertyId,
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:53]: Could these be separate to 1 per new line?
It improves visual readability and makes diffs simpler if items are added/removed.

File: source/UIAHandler/__init__.py [hunk 3]
```diff
@@ -49,6 +49,14 @@
 from . import remote as UIARemote
 
 
+baseCachePropertyIDs = set([
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:52]:
```suggestion
baseCachePropertyIDs = {
```

File: source/UIAHandler/__init__.py [hunk 4]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
 		UIA.IUIAutomationActiveTextPositionChangedEventHandler,
 	]
 
+	#: A cache of UIA notification kinds  to friendly names for logging
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:290]: Fix double space

```suggestion
	#: A cache of UIA notification kinds to friendly names for logging
```

File: source/UIAHandler/__init__.py [hunk 5]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		Generates a string representation of the given UIA notification kind,
+		suitable for logging.
+		This is the name part of the NotificationKind_* constant.
+		If a matching  constant can not be found,
+		then a string representation of the NotificationKind value itself is used.
+		E.g. "unknown notification kind 1234".
+		"""
+		name = self._notificationKindsToNamesCache.get(notificationKind)
+		if not name:
+			name = "unknown notification kind {notificationKind}"
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:308]: should this be an f-string? I cant't find a corresponding `.format`

```suggestion
			name = f"unknown notification kind {notificationKind}"
```

File: source/UIAHandler/__init__.py [hunk 6]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		for k, v in vars(UIA).items()
+		if k.startswith('NotificationKind_')
+	}
+
+	def getUIANotificationKindDebugString(self, notificationKind: int) -> str:
+		"""
+		Generates a string representation of the given UIA notification kind,
+		suitable for logging.
+		This is the name part of the NotificationKind_* constant.
+		If a matching  constant can not be found,
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:302]: Fix double space

```suggestion
		If a matching constant can not be found,
```

File: source/UIAHandler/__init__.py [hunk 7]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		for k, v in vars(UIA).items()
+		if k.startswith('NotificationProcessing_')
+	}
+
+	def getUIANotificationProcessingValueDebugString(self, notificationProcessing: int) -> str:
+		"""
+		Generates a string representation of the given UIA notification processing value,
+		suitable for logging.
+		This is the name part of the NotificationProcessing_* constant.
+		If a matching  constant can not be found,
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:323]: double space

```suggestion
		If a matching constant can not be found,
```

File: source/UIAHandler/__init__.py [hunk 8]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		then a string representation of the NotificationProcessing value itself is used.
+		E.g. "unknown notification processing value 1234".
+		"""
+		name = self._notificationProcessingValuesToNamesCache.get(notificationProcessing)
+		if not name:
+			name = "unknown notification processing value {notificationProcessing}"
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:329]: same here with f-string

File: source/UIAHandler/__init__.py [hunk 9]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		"""
+		Generates a string representation of the given UIA event ID,
+		suitable for logging.
+		This is the name part of the UIA_*EventId constant.
+		If a matching  constant can not be found, then a string representation of the ID itself is used.
+		E.g. "unknown event ID 1234".
+		"""
+		name = self._eventIDsToNamesCache.get(eventID)
+		if not name:
+			name = "unknown event ID {eventID}"
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:366]: same here with f-string

File: source/UIAHandler/__init__.py [hunk 10]
```diff
@@ -279,6 +287,123 @@ class UIAHandler(COMObject):
...
+		for k, v in vars(UIA).items()
+		if k.endswith('EventId')
+	}
+
+	def getUIAEventIDDebugString(self, eventID: int) -> str:
+		"""
+		Generates a string representation of the given UIA event ID,
+		suitable for logging.
+		This is the name part of the UIA_*EventId constant.
+		If a matching  constant can not be found, then a string representation of the ID itself is used.
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:361]: Double space

```suggestion
		If a matching constant can not be found, then a string representation of the ID itself is used.
```

File: source/UIAHandler/__init__.py [hunk 11]
```diff
@@ -572,9 +702,8 @@ def IUIAutomationEventHandler_HandleAutomationEvent(self,sender,eventID):
 			else:
 				if _isDebug():
 					log.debugWarning(
-						"HandleAutomationEvent: Dropping textChange event from element "
-						f"with class name {repr(sender.currentClassName)} "
-						f"and automation ID {repr(sender.CachedAutomationID)}"
+						f"HandleAutomationEvent: Dropping textChange event "
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:705]: Remove unnecessary f string

```suggestion
						"HandleAutomationEvent: Dropping textChange event "
```

File: source/UIAHandler/__init__.py [hunk 12]
```diff
@@ -612,26 +743,42 @@ def IUIAutomationEventHandler_HandleAutomationEvent(self,sender,eventID):
...
+			)
 		eventHandler.queueEvent(NVDAEventName,obj)
 
 	# The last UIAElement that received a UIA focus event
 	# This is updated no matter if this is a native element, the window is UIA blacklisted by NVDA, or  the element is proxied from MSAA 
 	lastFocusedUIAElement=None
 
 	def IUIAutomationFocusChangedEventHandler_HandleFocusChangedEvent(self,sender):
+		if _isDebug():
+			log.debug(f"handleFocusChangedEvent called with  element {self.getUIAElementDebugString(sender)}")
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:781]: Fix double space

```suggestion
			log.debug(f"handleFocusChangedEvent called with element {self.getUIAElementDebugString(sender)}")
```

File: source/UIAHandler/__init__.py [hunk 13]
```diff
@@ -722,6 +896,8 @@ def IUIAutomationPropertyChangedEventHandler_HandlePropertyChangedEvent(self,sen
 			isinstance(focus, NVDAObjects.UIA.UIA)
 			and self.clientObject.compareElements(focus.UIAElement, sender)
 		):
+			if _isDebug():
+				log.debug(f"propertyChange event is for  focus")
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:900]: Fix double space

```suggestion
				log.debug(f"propertyChange event is for focus")
```

File: source/UIAHandler/__init__.py [hunk 14]
```diff
@@ -790,13 +986,22 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(
 					f"activityId={activityId}"
 				)
 			return
+		if _isDebug():
+			log.debug(
+				f"Queuing  UIA_notification NVDA event "
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:991]: remove f string, fix double space

I'm not sure if this is an issue, but there are many strings that are unnecessarily marked as an f-string (i.e. no special formatting). 
I have seen reviews in the past encouraging f-strings only when required.

```suggestion
				"Queuing UIA_notification NVDA event "
```

File: source/UIAHandler/__init__.py [hunk 15]
```diff
@@ -904,23 +1141,49 @@ def _isUIAWindowHelper(self,hwnd):
...
+					return False
+			if isDebug:
+				log.debug("Treating as UIA")
+		else:
+			if isDebug:
+				log.debug("window does not have UIA server side provider. Treating as non-UIA")
 		return bool(res)
 
-	def isUIAWindow(self,hwnd):
+	def isUIAWindow(self, hwnd: int, isDebug=False) -> bool:
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:1159]: Add extra type information.

```suggestion
	def isUIAWindow(self, hwnd: int, isDebug: bool = False) -> bool:
```

**Comment by seanbudd** [comment]
Mostly trivial changes remaining

File: source/UIAHandler/__init__.py (hunk 8 above)
Comment (michaelDCurran) [reply source/UIAHandler/__init__.py:329]:
```suggestion
			name = f"unknown notification processing value {notificationProcessing}"
```

File: source/UIAHandler/__init__.py (hunk 9 above)
Comment (michaelDCurran) [reply source/UIAHandler/__init__.py:366]:
```suggestion
			name = f"unknown event ID {eventID}"
```

File: source/NVDAObjects/UIA/__init__.py (hunk 1 above)
Comment (seanbudd) [reply source/NVDAObjects/UIA/__init__.py:1307]: this is still an issue, same with the unnecessary f-strings.
However, as the remaining changes are trivial, I am approving in the mean-time

**Review by seanbudd — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@seanbudd  I've changed all  unnecessary f-strings to regular strings, and also added a missing  isDebug argument to the call to _isUIAWindowHelper. So another review please.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**